File: framework/src/play/src/main/scala/play/api/http/HttpEntity.scala
/*
 * Copyright (C) 2009-2018 Lightbend Inc. <https://www.lightbend.com>
 */

package play.api.http

import akka.stream.Materializer
import akka.stream.scaladsl.Source
import akka.util.ByteString
import play.api.mvc.Headers
import play.http.{ HttpEntity => JHttpEntity }

import scala.compat.java8.OptionConverters
import scala.concurrent.Future

/**
 * An HTTP entity.
 *
 * HTTP entities come in three flavors, [[HttpEntity.Strict]], [[HttpEntity.Streamed]] and [[HttpEntity.Chunked]].
 */
sealed trait HttpEntity {

  /**
   * The content type of the entity, if known.
   */
  def contentType: Option[String]

  /**
   * Whether it is known if this entity is empty or not.
   *
   * If this returns true, then the entity is definitely empty. If it returns false, the entity may or may not be empty.
   */
  def isKnownEmpty: Boolean

  /**
   * The content length of the entity, if known.
   */
  def contentLength: Option[Long]

  /**
   * The entity as a data stream.
   */
  def dataStream: Source[ByteString, _]

  /**
   * Consume the data from this entity.
   */
  def consumeData(implicit mat: Materializer): Future[ByteString] = {
    dataStream.runFold(ByteString.empty)(_ ++ _)
  }

  /**
   * Convert this entity to its Java counterpart.
   */
  def asJava: JHttpEntity

  /**
   * Return this entity as the given content type.
   */
  def as(contentType: String): HttpEntity
}

object HttpEntity {

  /**
   * No entity.
   */
  val NoEntity = Strict(ByteString.empty, None)

  /**
   * A strict entity.
   *
   * Strict entities are contained entirely in memory.
   *
   * @param data The data contained within this entity.
   * @param contentType The content type, if known.
   */
  final case class Strict(data: ByteString, contentType: Option[String]) extends HttpEntity {
    def isKnownEmpty = data.isEmpty
    def contentLength = Some(data.size)
    def dataStream = if (data.isEmpty) Source.empty[ByteString] else Source.single(data)
    override def consumeData(implicit mat: Materializer) = Future.successful(data)
    def asJava = new JHttpEntity.Strict(data, OptionConverters.toJava(contentType))
    def as(contentType: String) = copy(contentType = Some(contentType))
  }

  /**
   * A streamed entity.
   *
   * @param data The stream of data for this entity.
   * @param contentLength The content length, if known. If no content length is set, then this entity will be close
   *                      delimited.
   * @param contentType The content type, if known.
   */
  final case class Streamed(data: Source[ByteString, _], contentLength: Option[Long], contentType: Option[String]) extends HttpEntity {
    def isKnownEmpty = false
    def dataStream = data
    def asJava = new JHttpEntity.Streamed(
      data.asJava,
      OptionConverters.toJava(contentLength.asInstanceOf[Option[java.lang.Long]]),
      OptionConverters.toJava(contentType))
    def as(contentType: String) = copy(contentType = Some(contentType))
  }

  /**
   * A chunked entity.
   *
   * @param chunks The stream of chunks for this entity. Must be zero or more [[HttpChunk.Chunk]] elements, followed
   *               by zero or one [[HttpChunk.LastChunk]] elements. Any elements after the [[HttpChunk.LastChunk]]
   *               element will be ignored. If no [[HttpChunk.LastChunk]] element is sent, then the last chunk will
   *               contain no trailers.
   * @param contentType The content type, if known.
   */
  final case class Chunked(chunks: Source[HttpChunk, _], contentType: Option[String]) extends HttpEntity {
    def isKnownEmpty = false
    def contentLength = None
    def dataStream = chunks.collect {
      case HttpChunk.Chunk(data) => data
    }
    def asJava = new JHttpEntity.Chunked(chunks.asJava, OptionConverters.toJava(contentType))
    def as(contentType: String) = copy(contentType = Some(contentType))
  }
}

/**
 * An HTTP chunk.
 *
 * May either be a [[HttpChunk.Chunk]] containing data, or a [[HttpChunk.LastChunk]], signifying the last chunk in
 * a stream, optionally with trailing headers.
 */
sealed trait HttpChunk {

}

object HttpChunk {

  /**
   * A chunk.
   *
   * @param data The data for the chunk.
   */
  final case class Chunk(data: ByteString) extends HttpChunk {
    assert(data.nonEmpty, "Http chunks must not be empty")
  }

  /**
   * The last chunk.
   *
   * @param trailers The trailers.
   */
  final case class LastChunk(trailers: Headers) extends HttpChunk
}

File: framework/src/play/src/main/java/play/http/HttpEntity.java
/*
 * Copyright (C) 2009-2018 Lightbend Inc. <https://www.lightbend.com>
 */

package play.http;

import akka.japi.pf.PFBuilder;
import akka.stream.Materializer;
import akka.stream.javadsl.Source;
import akka.util.ByteString;
import play.api.http.HttpChunk;
import play.twirl.api.Content;
import play.twirl.api.Xml;
import scala.compat.java8.OptionConverters;

import java.util.Optional;
import java.util.concurrent.CompletionStage;

/**
 * An HTTP entity
 */
public abstract class HttpEntity {

    // sealed
    private HttpEntity() {}

    /**
     * @return The content type, if defined
     */
    public abstract Optional<String> contentType();

    /**
     * @return  Whether the entity is known to be empty or not.
     */
    public abstract boolean isKnownEmpty();

    /**
     * @return The content length, if known
     */
    public abstract Optional<Long> contentLength();

    /**
     * @return The stream of data.
     */
    public abstract Source<ByteString, ?> dataStream();

    /**
     * @param contentType    the content type to use, i.e. "text/html".
     * @return Return the entity as the given content type.
     */
    public abstract HttpEntity as(String contentType);

    /**
     * Consumes the data.
     *
     * This method should be used carefully, since if the source represents an ephemeral stream, then the entity may
     * not be usable after this method is invoked.
     * @param mat    the application's materializer.
     * @return a CompletionStage holding the data
     */
    public CompletionStage<ByteString> consumeData(Materializer mat) {
        return dataStream().runFold(ByteString.empty(), ByteString::concat, mat);
    }

    public abstract play.api.http.HttpEntity asScala();

    /**
     * No entity.
     */
    public static final HttpEntity NO_ENTITY = new Strict(ByteString.empty(), Optional.empty());

    /**
     * Create an entity from the given content.
     *
     * @param content The content.
     * @param charset The charset.
     *
     * @return the HTTP entity.
     */
    public static final HttpEntity fromContent(Content content, String charset) {
        String body;
        if (content instanceof Xml) {
            // See https://github.com/playframework/playframework/issues/2770
            body = content.body().trim();
        } else {
            body = content.body();
        }
        return new Strict(ByteString.fromString(body, charset), Optional.of(content.contentType() + "; charset=" + charset));
    }

    /**
     * Create an entity from the given String.
     *
     * @param content The content.
     * @param charset The charset.
     * @return the HTTP entity.
     */
    public static final HttpEntity fromString(String content, String charset) {
        return new Strict(ByteString.fromString(content, charset), Optional.of("text/plain; charset=" + charset));
    }

    /**
     * Convert the given source of ByteStrings to a chunked entity.
     *
     * @param data The source.
     * @param contentType The optional content type.
     * @return The ByteStrings.
     */
    public static final HttpEntity chunked(Source<ByteString, ?> data, Optional<String> contentType) {
        return new Chunked(data.map(HttpChunk.Chunk::new), contentType);
    }

    /**
     * A strict entity, where all the data for it is in memory.
     */
    public final static class Strict extends HttpEntity {
        private final ByteString data;
        private final Optional<String> contentType;

        public Strict(ByteString data, Optional<String> contentType) {
            this.data = data;
            this.contentType = contentType;
        }

        public ByteString data() {
            return data;
        }

        @Override
        public Optional<String> contentType() {
            return contentType;
        }

        @Override
        public boolean isKnownEmpty() {
            return data.isEmpty();
        }

        @Override
        public Optional<Long> contentLength() {
            return Optional.of((long) data.length());
        }

        @Override
        public HttpEntity as(String contentType) {
            return new Strict(data, Optional.of(contentType));
        }

        @Override
        public Source<ByteString, ?> dataStream() {
            return Source.<ByteString>single(data);
        }

        @Override
        public play.api.http.HttpEntity asScala() {
            return new play.api.http.HttpEntity.Strict(data, OptionConverters.toScala(contentType));
        }
    }

    /**
     * A streamed entity, backed by a source.
     */
    public final static class Streamed extends HttpEntity {
        private final Source<ByteString, ?> data;
        private final Optional<Long> contentLength;
        private final Optional<String> contentType;

        public Streamed(Source<ByteString, ?> data, Optional<Long> contentLength, Optional<String> contentType) {
            this.data = data;
            this.contentType = contentType;
            this.contentLength = contentLength;
        }

        public Source<ByteString, ?> data() {
            return data;
        }

        @Override
        public Optional<String> contentType() {
            return contentType;
        }

        @Override
        public boolean isKnownEmpty() {
            return false;
        }

        @Override
        public Optional<Long> contentLength() {
            return contentLength;
        }

        @Override
        public HttpEntity as(String contentType) {
            return new Streamed(data, contentLength, Optional.of(contentType));
        }

        @Override
        public Source<ByteString, ?> dataStream() {
            return data;
        }

        @Override
        @SuppressWarnings("unchecked")
        public play.api.http.HttpEntity asScala() {
            return new play.api.http.HttpEntity.Streamed(data.asScala(),
                    /* scala Option[Long] produces a Java generic signature of Option<Object>, so we need to do an
                       unchecked cast here to get it to typecheck */
                    (scala.Option) OptionConverters.toScala(contentLength),
                    OptionConverters.toScala(contentType));
        }
    }

    /**
     * A chunked entity, backed by a source of chunks.
     */
    public final static class Chunked extends HttpEntity {
        private final Source<HttpChunk, ?> chunks;
        private final Optional<String> contentType;

        public Chunked(Source<HttpChunk, ?> chunks, Optional<String> contentType) {
            this.chunks = chunks;
            this.contentType = contentType;
        }

        public Source<HttpChunk, ?> chunks() {
            return chunks;
        }

        @Override
        public Optional<String> contentType() {
            return contentType;
        }

        @Override
        public boolean isKnownEmpty() {
            return false;
        }

        @Override
        public Optional<Long> contentLength() {
            return Optional.empty();
        }

        @Override
        public HttpEntity as(String contentType) {
            return new Chunked(chunks, Optional.of(contentType));
        }

        @Override
        public Source<ByteString, ?> dataStream() {
            return chunks.<ByteString>collect(new PFBuilder<HttpChunk, ByteString>()
                    .match(HttpChunk.Chunk.class, HttpChunk.Chunk::data)
                    .build()
            );
        }

        @Override
        public play.api.http.HttpEntity asScala() {
            return new play.api.http.HttpEntity.Chunked(chunks.asScala(), OptionConverters.toScala(contentType));
        }
    }
}
